microceph/ceph/osd.go

```diff
@@ -438,28 +439,413 @@ func AddOSD(s *state.State, path string, wipe bool, encrypt bool) error {
+		// e.g. if the device is broken, we still want to remove it from the cluster
+		logger.Warnf("Fault during device wipe: %v", err)
+	}
+	return nil
+}
+
+// timeoutWipe wipes the given device with a timeout, in order not to hang on broken disks
+func timeoutWipe(path string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+
+	_, err := processExec.RunCommandContext(
+		ctx,
+		"dd", "if=/dev/zero",
+		fmt.Sprintf("of=%s", path),
+		"bs=4M", "count=10", "status=none",
+	)
+	return err
+}
+
+func doRemoveOSD(ctx context.Context, s common.StateInterface, osd int64, bypassSafety bool) error {
+	var err error
+
+	// general sanity
+	err = sanityCheck(s, osd)
+	if err != nil {
+		return err
+	}
+
+	if !bypassSafety {
+		// check: at least 3 OSDs
+		err = checkMinOSDs(s, osd)
 		if err != nil {
-			return fmt.Errorf("Failed to fetch disks: %w", err)
+			return err
+		}
+	}
+
+	err = scaleDownFailureDomain(s, osd)
+	if err != nil {
+		return err
+	}
+
+	// check if the osd is still in the cluster -- if we're being re-run, it might not be
+	isPresent, err := haveOSDInCeph(osd)
+	if err != nil {
+		return fmt.Errorf("Failed to check if osd.%d is present in Ceph: %w", osd, err)
+	}
+	// reweight/drain data
+	if isPresent {
+		reweightOSD(ctx, osd, 0)
+	}
+	// perform safety check for stopping
+	if isPresent && !bypassSafety {
+		err = safetyCheckStop(osd)
+		if err != nil {
+			return err
+		}
+	}
+	// take the OSD out and down
+	if isPresent {
+		err = outDownOSD(osd)
+		if err != nil {
+			return err
+		}
+	}
+	// stop the OSD service
+	if isPresent {
+		err = killOSD(osd)
+	}
+	if err != nil {
+		return err
+	}
+	// perform safety check for destroying
+	if isPresent && !bypassSafety {
+		err = safetyCheckDestroy(osd)
+		if err != nil {
+			return err
+		}
+	}
+	// purge the OSD
+	if isPresent {
+		err = purgeOSD(osd)
+		if err != nil {
+			return err
 		}
+	}
+	// Wipe the underlying blocking device
+	err = wipeDevice(s, osd)
+	if err != nil {
+		return err
+	}
+	// Remove osd config
+	err = removeOSDConfig(osd)
+	if err != nil {
+		return err
+	}
+	// Remove db entry
+	err = database.OSDQuery.Delete(s.ClusterState(), osd)
+	if err != nil {
+		logger.Errorf("Failed to remove osd.%d from database: %v", osd, err)
+		return fmt.Errorf("Failed to remove osd.%d from database: %w", osd, err)
+	}
+	return nil
+}
 
-		for _, disk := range records {
-			disks = append(disks, types.Disk{
-				Location: disk.Member,
-				OSD:      int64(disk.OSD),
-				Path:     disk.Path,
-			})
+func checkMinOSDs(s common.StateInterface, osd int64) error {
+	// check if we have at least 3 OSDs post-removal
+	disks, err := database.OSDQuery.List(s.ClusterState())
+	if err != nil {
+		return err
+	}
+	if len(disks) <= 3 {
+		return fmt.Errorf("Cannot remove osd.%d we need at least 3 OSDs, have %d", osd, len(disks))
+	}
+	return nil
+}
+
+func outDownOSD(osd int64) error {
```

Comment:
Is this function necessary? I think if you reweight the OSD to 0 and `ok-to-stop` and `safe-to-destroy` return true, then the OSD can be safely removed; it doesn't need to be marked as out, as far as I recall. At least, that's how we do it in the OSD charm.

Reply:
In my testing the OSD removal was much more robust if you took out the OSD explicitly (we might want to consider doing it for the charm as well)